### PR TITLE
Emv4

### PR DESCRIPTION
**Pull request by merlokk**
Added SDA, DDA, fast DDA, CDA calculations to `hf emv exec`

### CODE REVIEW

**Comment by merlokk** [comment]
ready to merge.
before merge (after code review)))) ) it will needs to delete travis files.

**Comment by iceman1001** [comment]
Impressive work I say.   But Monster merge...  
This PR could have been split up in many minor ones,  which would have been much easier to verify.
like the crypto change,   the new crypto parts,   
why do you add the test-cases for the crypto?

**Comment by merlokk** [comment]
because anyway some of test lays in polarssl files. and because i need to check that cryptosystem works....  There is huge hill of crypto code and I need to be sure that crypto works to debug it.

**Comment by iceman1001** [comment]
I'm not going to merge this one.  This one might take some months to become merged because of size.
No time to test it.   I leave this one for the others.

Keep up the great work!

**Comment by merlokk** [comment]
Crypto part works and tested by command. and the rest - porting of emv_tools library and 
my code laid info `emvcore` and `emvcmd` and it not a big part.

**Comment by iceman1001** [comment]
As mentioned,  I leave this one for the others.

**Comment by pwpiwi** [comment]
"the others" ?  :smile:

**Comment by iceman1001** [comment]
well, we do have more maintainers last time I checked.

**Comment by merlokk** [comment]
There are 42 files changed in this PR
lets look:

these files we know)

1. CHANGELOG.md
2. client/Makefile
3. client/obj/emv/test/.dummy

these must be deleted after review)

4. .travis.yml
5. travis_test_commands.scr

these are tests for crypto parts. almost all of them from `emv_tools` library.
This part tests with `hf emv test`

6. client/emv/test/cda_test.c
7. client/emv/test/cda_test.h
8. client/emv/test/crypto_test.c
9. client/emv/test/crypto_test.h
10. client/emv/test/cryptotest.c
11. client/emv/test/cryptotest.h
12. client/emv/test/dda_test.c
13. client/emv/test/dda_test.h
14. client/emv/test/sda_test.c
15. client/emv/test/sda_test.h
16. client/scripting.c

these are `polarssl` crypto library
This part tests with `hf emv test`

17. common/polarssl/aes.h
18. common/polarssl/bignum.c
19. common/polarssl/bignum.h
20. common/polarssl/bn_mul.h
21. common/polarssl/des.c
22. common/polarssl/polarssl_config.h
23. common/polarssl/rsa.c
24. common/polarssl/rsa.h
25. common/sha1.c > common/polarssl/sha1.c
26. common/sha1.h > common/polarssl/sha1.h

these are from `emv_tools` library

27. client/emv/capk.txt
28. client/emv/crypto.c - tests with `hf emv test`
29. client/emv/crypto.h - tests with `hf emv test`
30. client/emv/crypto_libgcrypt.c - not needs. will be deleted after this message
31. client/emv/crypto_polarssl.c - this mine and it tests with `hf emv test`
32. client/emv/emv_pk.c
33. client/emv/emv_pk.h
34. client/emv/emv_pki.c
35. client/emv/emv_pki.h
36. client/emv/emv_pki_priv.c
37. client/emv/emv_pki_priv.h
38. client/emv/emv_tags.c - here just some changes
39. client/emv/crypto_backend.h - tests with `hf emv test`

these files with my changes

40. client/emv/cmdemv.c
41. client/emv/emvcore.c
42. client/emv/emvcore.h

so)
in this PR we have 3 changed files and many ported files from `polarssl` and `emv_tools`. Almost all ported files can be checked with `hf emv test` command.

**Comment by merlokk** [comment]
Lets look and merge)
or wait to stabilize release?

**Comment by pwpiwi** [comment]
I may find some time over the upcoming weekend to have a look at your code.

File: client/emv/cmdemv.c [hunk 1]
```diff
@@ -294,19 +296,23 @@ int UsageCmdHFEMVExec(void) {
 }
 
 #define TLV_ADD(tag, value)( tlvdb_add(tlvRoot, tlvdb_fixed(tag, sizeof(value) - 1, (const unsigned char *)value)) )
+#define dreturn(n) do{DropField();return n;}while(0)
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: do ... while(0)? Doesn't it work without?

File: client/emv/cmdemv.c [hunk 2]
```diff
@@ -462,26 +477,26 @@ int CmdHFEMVExec(const char *cmd) {
...
+		dreturn(4);
 	}
 	PrintAndLog("PDOL data[%d]: %s", pdol_data_tlv_data_len, sprint_hex(pdol_data_tlv_data, pdol_data_tlv_data_len));
 
 	PrintAndLog("\n* GPO.");
 	res = EMVGPO(true, pdol_data_tlv_data, pdol_data_tlv_data_len, buf, sizeof(buf), &len, &sw, tlvRoot);
 	
 	free(pdol_data_tlv_data);
-	free(pdol_data_tlv);
+	//free(pdol_data_tlv); --- free on exit.
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: But it is not freed by dreturn().   tlvdbfree required as well on dreturn()?  ???

**Comment by pwpiwi** [comment]
```
Testing key length 1024 passed. (1018 ms)
Testing key length 1152 passed. (137 ms)
Testing key length 1408 passed. (236 ms)
Testing key length 1984 passed. (10276 ms)
Testing key length 2048 passed. (1511 ms)
Testing key length 3072 passed. (26111 ms)
Testing key length 4096 passed. (39510 ms)
```
I would have expected monotonic increasing execution times. Is there an explanation for this behaviour?

**Comment by merlokk** [comment]
Yes, creating key takes random time. And it May be very strange, but its ok.

File: client/emv/cmdemv.c (hunk 1 above)
Comment (iceman1001) [reply client/emv/cmdemv.c]: I belive @pwpiwi  means this:

 #define dreturn()  { DropField(); }

**Comment by pwpiwi** [comment]
Wrong samples in `hf emv exec` help:
```
Samples:
 hf emv pse -s -> select card
 hf emv pse -s -t -a -> select card, show responses in TLV, show APDU
```

**Comment by merlokk** [comment]
Ill delete travis before merge.

**Comment by pwpiwi** [comment]
Usage incomplete in `hf mf emv` help:
```
Usage:  hf emv exec [-s][-a][-t]
```

**Comment by pwpiwi** [comment]
`hf mf emv -s -v` with a VISA card:
```
* DDA
CA public key index 0x8
fopen: No such file or directory
ERROR: Key not found. Exit.
```
Is this expected?

**Comment by merlokk** [comment]
No, it needs to have emv/capk. txt with Public keys

**Comment by merlokk** [comment]
Ill fix in 2-3 hours

File: client/emv/cmdemv.c (hunk 1 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: works. )   cool)

File: client/emv/cmdemv.c (hunk 2 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: fixed. i plain to do some refactoring before making card scan

**Comment by merlokk** [comment]
done. do you managed to make DDA? in emv/capk.txt there is a 0x08 key

**Comment by merlokk** [comment]
changed some travis behavior. it not change PR because i will delete travis files after review done. 
P.S. have you managed to check?

**Comment by pwpiwi** [comment]
I cannot see any change to capk.txt handling? Why should it work now?

**Comment by merlokk** [comment]
what OS do you use?
do you launch proxmark3 from place in sources? like `pm3/client/proxmark3`
Do you have this file in `pm3/client/emv/capk.txt`

as i see it already there.

**Comment by merlokk** [comment]
it loads from here
https://github.com/Proxmark/proxmark3/pull/512/files#diff-72476a17bcbfdca6d67c0ed2bfc7b61dR492

**Comment by pwpiwi** [comment]
Oh. You are assuming that you start the client from the client directory. I don't (I use `client\proxmark3` to start the client. See the file handling for lua scripts and hardnested files how to make it more flexible.

**Comment by merlokk** [comment]
done

**Comment by pwpiwi** [comment]
Is this intended?
https://github.com/merlokk/proxmark3/blob/179f2904d03d8a6562e1ef5a357c462db90ddee7/client/emv/emvcore.c#L209
Or should it be filled with '\0' (marking the end of the string) instead?

File: client/emv/emvcore.c [hunk 3]
```diff
@@ -464,32 +464,330 @@ int EMVGenerateChallenge(bool LeaveFieldON, uint8_t *Result, size_t MaxResultLen
...
+			return 5;
+		}
+
+		PrintAndLog("DDOL data[%d]: %s", ddol_data_tlv->len, sprint_hex(ddol_data_tlv->value, ddol_data_tlv->len));
+
+		PrintAndLog("\n* Internal Authenticate");
+		int res = EMVInternalAuthenticate(true, (uint8_t *)ddol_data_tlv->value, ddol_data_tlv->len, buf, sizeof(buf), &len, &sw, NULL);
+		if (res) {	
+			PrintAndLog("Internal Authenticate error(%d): %4x. Exit...", res, sw);
+			return 6;
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: need to free ` ddol_data_tlv` ?

File: client/emv/cmdemv.c [hunk 4]
```diff
@@ -753,7 +798,7 @@ int CmdHFEMVExec(const char *cmd) {
 				res = MSCComputeCryptoChecksum(true, (uint8_t *)udol_data_tlv->value, udol_data_tlv->len, buf, sizeof(buf), &len, &sw, tlvRoot);
 				if (res) {
 					PrintAndLog("ERROR Compute Crypto Checksum. APDU error %4x", sw);
-					return 5;
+					dreturn(9);
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Need to free `udol_data_tlv` ?

File: client/emv/crypto_polarssl.c [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+	else
+		return 0;
+}
+
+static struct crypto_hash *crypto_hash_polarssl_open(enum crypto_algo_hash hash)
+{
+	struct crypto_hash_polarssl *ch = malloc(sizeof(*ch));
+
+	if (hash != HASH_SHA_1)
+		return NULL;
```
Comment (pwpiwi) [inline client/emv/crypto_polarssl.c]: Need to free `ch`.

File: client/emv/emvcore.c [hunk 6]
```diff
@@ -464,32 +464,330 @@ int EMVGenerateChallenge(bool LeaveFieldON, uint8_t *Result, size_t MaxResultLen
...
+	const struct tlv *sda_tlv = tlvdb_get(tlv, 0x21, NULL);
+	if (!sda_tlv || sda_tlv->len < 1) {
+		PrintAndLog("ERROR: Can't find input list for Offline Data Authentication. Exit.");
+		return 3;
+	}
+
+	struct emv_pk *issuer_pk = emv_pki_recover_issuer_cert(pk, tlv);
+	if (!issuer_pk) {
+		PrintAndLog("ERROR: Issuer certificate not found. Exit.");
+		return 2;
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Need to free `pk` ?

File: client/emv/emvcore.c [hunk 7]
```diff
@@ -464,32 +464,330 @@ int EMVGenerateChallenge(bool LeaveFieldON, uint8_t *Result, size_t MaxResultLen
...
+	}
 	
-	// Processing of Issuer Public Key Certificate (0x90)
-	//Certificate = 
+	struct emv_pk *issuer_pk = emv_pki_recover_issuer_cert(pk, tlv);
+	if (!issuer_pk) {
+		PrintAndLog("ERROR: Issuer certificate not found. Exit.");
+		return 2;
+	}
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Need to free `pk` ?

File: client/emv/emvcore.c [hunk 8]
```diff
@@ -464,32 +464,330 @@ int EMVGenerateChallenge(bool LeaveFieldON, uint8_t *Result, size_t MaxResultLen
...
+	const struct tlv *sda_tlv = tlvdb_get(tlv, 0x21, NULL);
+	if (!sda_tlv || sda_tlv->len < 1) {
+		PrintAndLog("ERROR: Can't find input list for Offline Data Authentication. Exit.");
+		return 3;
+	}
+
+	struct emv_pk *issuer_pk = emv_pki_recover_issuer_cert(pk, tlv);
+	if (!issuer_pk) {
+		PrintAndLog("ERROR: Issuer certificate not found. Exit.");
+		return 2;
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Need to free `pk` ?

File: client/emv/emv_pki_priv.c [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+	pk->hash_algo = hash_algo;
+	memcpy(pk->modulus, mod, modlen);
+	memcpy(pk->exp, exp, explen);
+
+	free(mod);
+	free(exp);
+
+	struct crypto_hash *ch = crypto_hash_open(pk->hash_algo);
+	if (!ch)
+		return false;
```
Comment (pwpiwi) [inline client/emv/emv_pki_priv.c]: Need to free `pk` ?

**Comment by pwpiwi** [comment]
https://scan.coverity.com   :smiley:

**Comment by iceman1001** [comment]
coverty scan is really good at getting at ironing out these kinds of bugs.

**Comment by merlokk** [comment]
>Is this intended?
yes, there is a 32 hex symbols and some of them filled with data

**Comment by pwpiwi** [comment]
Then why is it an array if 100 chars?

**Comment by merlokk** [comment]
it because there may be another cryptogram with bigger length. It made for not to forget this.  truncate it?

**Comment by pwpiwi** [comment]
I was just curious..

File: client/emv/cmdemv.c (hunk 4 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: deleted

**Comment by merlokk** [comment]
checked. works.

**Comment by pwpiwi** [comment]
This generates a warning because it is never used:
https://github.com/merlokk/proxmark3/blob/899b27aa06b762a4b0218fda06518c4fadcea508/client/emv/emvcore.c#L19
Please make it a comment or remove.

File: client/emv/cmdemv.c (hunk 4 above)
Comment (pwpiwi) [reply client/emv/cmdemv.c]: Deleted? What?

Comment (merlokk) [reply client/emv/cmdemv.c]: deleted `udol_data_tlv` declaration and move it to another place

Comment (merlokk) [reply client/emv/cmdemv.c]: found. will fix

**Comment by merlokk** [comment]
done

**Comment by iceman1001** [comment]
What more is needed for this one to become merged?

**Comment by pwpiwi** [comment]
Only one test regarding the public keys.

**Comment by iceman1001** [comment]
and maybe the cryptolib should be under  common/  folder instead in emv/

**Comment by merlokk** [comment]
crypto library already in the separate folder. in emv folder there is emv-tools modified library. maybe i can move emv-tools to separate folder under emv folder. is it makes sense?

**Comment by pwpiwi** [comment]
hf emv exec -s -v with VISA card now works (well, similar to @iceman1001's comments: produces a lot of output which I don't really understand :smile: )

What you need do to before merge:
* Raise another PR to add emv/capk.txt to homebrew
* remove the travis stuff

**Comment by merlokk** [comment]
https://github.com/Proxmark/homebrew-proxmark3/pull/16

done

**Pull request merged**